File: test/analyze-stats.py
import json
import os
from collections import defaultdict

def analyze_build_stats():
    file_path = 'build-stats.json'
    
    print("Starting analysis of build-stats.json...")
    print(f"File size: {os.path.getsize(file_path) / (1024*1024):.2f} MB")
    
    try:
        print("Loading JSON file...")
        with open(file_path, 'r', encoding='utf-8') as f:
            stats = json.load(f)
        
        print("\n=== WEBPACK BUILD STATS ANALYSIS ===\n")
        
        # Basic info
        print(f"Webpack Version: {stats.get('version', 'Unknown')}")
        print(f"Build Time: {stats.get('time', 'Unknown')} ms")
        if 'builtAt' in stats:
            from datetime import datetime
            built_at = datetime.fromtimestamp(stats['builtAt'] / 1000)
            print(f"Built At: {built_at.isoformat()}")
        
        # Assets analysis
        if 'assets' in stats and isinstance(stats['assets'], list):
            print(f"\n=== ASSETS ANALYSIS ===")
            print(f"Total Assets: {len(stats['assets'])}")
            
            # Sort by size
            assets_with_size = [asset for asset in stats['assets'] if 'size' in asset]
            sorted_assets = sorted(assets_with_size, key=lambda x: x['size'], reverse=True)[:20]
            
            print(f"\nTop 20 Largest Assets:")
            for i, asset in enumerate(sorted_assets, 1):
                size_kb = asset['size'] / 1024
                size_mb = size_kb / 1024
                size_display = f"{size_mb:.2f}MB" if size_mb > 1 else f"{size_kb:.2f}KB"
                print(f"{i}. {asset['name']} - {size_display}")
            
            # Total bundle size
            total_size = sum(asset.get('size', 0) for asset in stats['assets'])
            total_size_mb = total_size / (1024 * 1024)
            print(f"\nTotal Bundle Size: {total_size_mb:.2f}MB")
        
        # Modules analysis
        if 'modules' in stats and isinstance(stats['modules'], list):
            print(f"\n=== MODULES ANALYSIS ===")
            print(f"Total Modules: {len(stats['modules'])}")
            
            # Sort modules by size
            modules_with_size = [module for module in stats['modules'] if 'size' in module]
            sorted_modules = sorted(modules_with_size, key=lambda x: x['size'], reverse=True)[:20]
            
            print(f"\nTop 20 Largest Modules:")
            for i, module in enumerate(sorted_modules, 1):
                size_kb = module['size'] / 1024
                name = module.get('name') or module.get('identifier', 'Unknown module')
                # Truncate long names
                display_name = name[:80] + '...' if len(name) > 80 else name
                print(f"{i}. {display_name} - {size_kb:.2f}KB")
            
            # NPM packages analysis
            package_sizes = defaultdict(int)
            for module in stats['modules']:
                name = module.get('name') or module.get('identifier', '')
                if 'node_modules' in name:
                    # Extract package name
                    parts = name.split('node_modules/')
                    if len(parts) > 1:
                        package_part = parts[-1].split('/')[0]
                        if package_part.startswith('@'):
                            # Scoped package
                            package_parts = parts[-1].split('/')[:2]
                            package_name = '/'.join(package_parts)
                        else:
                            package_name = package_part
                        package_sizes[package_name] += module.get('size', 0)
            
            if package_sizes:
                top_packages = sorted(package_sizes.items(), key=lambda x: x[1], reverse=True)[:15]
                print(f"\nTop 15 Largest NPM Packages:")
                for i, (package_name, size) in enumerate(top_packages, 1):
                    size_kb = size / 1024
                    size_mb = size_kb / 1024
                    size_display = f"{size_mb:.2f}MB" if size_mb > 1 else f"{size_kb:.2f}KB"
                    print(f"{i}. {package_name} - {size_display}")
        
        # Chunks analysis
        if 'chunks' in stats and isinstance(stats['chunks'], list):
            print(f"\n=== CHUNKS ANALYSIS ===")
            print(f"Total Chunks: {len(stats['chunks'])}")
            
            chunks_with_size = [chunk for chunk in stats['chunks'] if 'size' in chunk]
            sorted_chunks = sorted(chunks_with_size, key=lambda x: x['size'], reverse=True)[:10]
            
            print(f"\nTop 10 Largest Chunks:")
            for i, chunk in enumerate(sorted_chunks, 1):
                size_kb = chunk['size'] / 1024
                size_mb = size_kb / 1024
                size_display = f"{size_mb:.2f}MB" if size_mb > 1 else f"{size_kb:.2f}KB"
                names = ', '.join(chunk.get('names', [])) if chunk.get('names') else f"Chunk {chunk.get('id', 'Unknown')}"
                print(f"{i}. {names} - {size_display}")
        
        # Build status
        print(f"\n=== BUILD STATUS ===")
        print(f"Errors: {stats.get('errorsCount', 0)}")
        print(f"Warnings: {stats.get('warningsCount', 0)}")
        
        if stats.get('errors') and len(stats['errors']) > 0:
            print(f"\nFirst 3 Errors:")
            for i, error in enumerate(stats['errors'][:3], 1):
                print(f"{i}. {str(error)[:200]}...")
        
        if stats.get('warnings') and len(stats['warnings']) > 0:
            print(f"\nFirst 3 Warnings:")
            for i, warning in enumerate(stats['warnings'][:3], 1):
                print(f"{i}. {str(warning)[:200]}...")
        
        print(f"\n=== OPTIMIZATION RECOMMENDATIONS ===")
        print("1. Check the largest assets and modules above for optimization opportunities")
        print("2. Consider code splitting for large chunks")
        print("3. Review NPM packages - consider lighter alternatives for the largest ones")
        print("4. Use webpack-bundle-analyzer for visual analysis")
        print("5. Consider dynamic imports for non-critical modules")
        print("6. Enable tree shaking for unused code elimination")
        print("7. Consider using Module Federation for microfrontend optimization")
        
    except Exception as e:
        print(f"Error analyzing build stats: {str(e)}")
        if 'JSON' in str(e):
            print("The file might be corrupted or not valid JSON")

if __name__ == "__main__":
    analyze_build_stats()

File: test/analyze-stats.js
const fs = require("fs");
const path = require("path");

console.log("Starting analysis of build-stats.json...");

try {
  // Read and parse the JSON file
  const statsPath = path.join(__dirname, "build-stats.json");
  console.log("Reading file:", statsPath);

  const statsContent = fs.readFileSync(statsPath, "utf8");
  console.log(
    "File size:",
    Math.round((statsContent.length / 1024 / 1024) * 100) / 100,
    "MB"
  );

  console.log("Parsing JSON...");
  const stats = JSON.parse(statsContent);

  console.log("\n=== WEBPACK BUILD STATS ANALYSIS ===\n");

  // Basic info
  console.log("Webpack Version:", stats.version || "Unknown");
  console.log("Build Time:", stats.time ? `${stats.time}ms` : "Unknown");
  console.log(
    "Built At:",
    stats.builtAt ? new Date(stats.builtAt).toISOString() : "Unknown"
  );

  // Assets analysis
  if (stats.assets && Array.isArray(stats.assets)) {
    console.log("\n=== ASSETS ANALYSIS ===");
    console.log("Total Assets:", stats.assets.length);

    // Sort by size
    const sortedAssets = stats.assets
      .filter((asset) => asset.size)
      .sort((a, b) => b.size - a.size)
      .slice(0, 20); // Top 20 largest assets

    console.log("\nTop 20 Largest Assets:");
    sortedAssets.forEach((asset, index) => {
      const sizeKB = Math.round((asset.size / 1024) * 100) / 100;
      const sizeMB = Math.round((asset.size / 1024 / 1024) * 100) / 100;
      const sizeDisplay = sizeMB > 1 ? `${sizeMB}MB` : `${sizeKB}KB`;
      console.log(`${index + 1}. ${asset.name} - ${sizeDisplay}`);
    });

    // Total bundle size
    const totalSize = stats.assets.reduce(
      (sum, asset) => sum + (asset.size || 0),
      0
    );
    const totalSizeMB = Math.round((totalSize / 1024 / 1024) * 100) / 100;
    console.log(`\nTotal Bundle Size: ${totalSizeMB}MB`);
  }

  // Modules analysis
  if (stats.modules && Array.isArray(stats.modules)) {
    console.log("\n=== MODULES ANALYSIS ===");
    console.log("Total Modules:", stats.modules.length);

    // Sort modules by size
    const sortedModules = stats.modules
      .filter((module) => module.size)
      .sort((a, b) => b.size - a.size)
      .slice(0, 20);

    console.log("\nTop 20 Largest Modules:");
    sortedModules.forEach((module, index) => {
      const sizeKB = Math.round((module.size / 1024) * 100) / 100;
      const name = module.name || module.identifier || "Unknown module";
      // Truncate long names
      const displayName =
        name.length > 80 ? name.substring(0, 80) + "..." : name;
      console.log(`${index + 1}. ${displayName} - ${sizeKB}KB`);
    });

    // Module types analysis
    const moduleTypes = {};
    stats.modules.forEach((module) => {
      const name = module.name || module.identifier || "";
      if (name.includes("node_modules")) {
        const match = name.match(/node_modules\/([^\/]+)/);
        if (match) {
          const packageName = match[1];
          moduleTypes[packageName] =
            (moduleTypes[packageName] || 0) + (module.size || 0);
        }
      }
    });

    const topPackages = Object.entries(moduleTypes)
      .sort(([, a], [, b]) => b - a)
      .slice(0, 15);

    if (topPackages.length > 0) {
      console.log("\nTop 15 Largest NPM Packages:");
      topPackages.forEach(([packageName, size], index) => {
        const sizeKB = Math.round((size / 1024) * 100) / 100;
        const sizeMB = Math.round((size / 1024 / 1024) * 100) / 100;
        const sizeDisplay = sizeMB > 1 ? `${sizeMB}MB` : `${sizeKB}KB`;
        console.log(`${index + 1}. ${packageName} - ${sizeDisplay}`);
      });
    }
  }

  // Chunks analysis
  if (stats.chunks && Array.isArray(stats.chunks)) {
    console.log("\n=== CHUNKS ANALYSIS ===");
    console.log("Total Chunks:", stats.chunks.length);

    const sortedChunks = stats.chunks
      .filter((chunk) => chunk.size)
      .sort((a, b) => b.size - a.size)
      .slice(0, 10);

    console.log("\nTop 10 Largest Chunks:");
    sortedChunks.forEach((chunk, index) => {
      const sizeKB = Math.round((chunk.size / 1024) * 100) / 100;
      const sizeMB = Math.round((chunk.size / 1024 / 1024) * 100) / 100;
      const sizeDisplay = sizeMB > 1 ? `${sizeMB}MB` : `${sizeKB}KB`;
      const names =
        chunk.names && chunk.names.length > 0
          ? chunk.names.join(", ")
          : `Chunk ${chunk.id}`;
      console.log(`${index + 1}. ${names} - ${sizeDisplay}`);
    });
  }

  // Errors and warnings
  console.log("\n=== BUILD STATUS ===");
  console.log("Errors:", stats.errorsCount || 0);
  console.log("Warnings:", stats.warningsCount || 0);

  if (stats.errors && stats.errors.length > 0) {
    console.log("\nErrors:");
    stats.errors.slice(0, 5).forEach((error, index) => {
      console.log(`${index + 1}. ${error}`);
    });
  }

  if (stats.warnings && stats.warnings.length > 0) {
    console.log("\nWarnings:");
    stats.warnings.slice(0, 5).forEach((warning, index) => {
      console.log(`${index + 1}. ${warning}`);
    });
  }

  console.log("\n=== OPTIMIZATION RECOMMENDATIONS ===");
  console.log(
    "1. Check the largest assets and modules above for optimization opportunities"
  );
  console.log("2. Consider code splitting for large chunks");
  console.log(
    "3. Review NPM packages - consider lighter alternatives for the largest ones"
  );
  console.log("4. Use webpack-bundle-analyzer for visual analysis");
  console.log("5. Consider dynamic imports for non-critical modules");
} catch (error) {
  console.error("Error analyzing build stats:", error.message);
  if (error.message.includes("JSON")) {
    console.error("The file might be corrupted or not valid JSON");
  }
}
